src-tauri/src/chaosdna.rs

```diff
@@ -0,0 +1,76 @@
+use rand::{Rng, distributions::{Distribution, Standard}};
+use crate::primer::Base;
+
+
+pub trait RandError {
+    fn modify_base(&self, input_base: &Base) -> Base; 
+    fn full_chaos(&self, initial_seq: &Vec<Base>, edit_prob: f64, mod_prob: f64) -> Vec<Base>;
+}
+
+
+pub struct Chaos {
+}
+
+
+impl RandError for Chaos {
+    // Takes a base vector and probability as input, outputs a base vector with modified bases
+    fn modify_base(&self, input_base: &Base) -> Base {
+	let mut rng = rand::thread_rng();
+	loop {
+	    let new_base = rng.gen::<Base>();
+	    if new_base != *input_base {
+		return new_base;
+	    }
+	}
+    }
+
+
+    fn full_chaos(&self, initial_seq: &Vec<Base>, edit_prob: f64, del_prob: f64) -> Vec<Base> {
+	let mut rng = rand::thread_rng();
+	initial_seq
+	    .iter()
+	    .filter_map(|base| {
+		if rng.gen_bool(edit_prob) {
+		    if rng.gen_bool(del_prob) {
+			None
+		    }
+		    else {
+			Some(self.modify_base(base))
+		    }
+		}
+		else {
+		    Some(*base)
+		}
+	    })
+	    .collect()
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::primer::Base;
+
+    #[test]
+    fn test_mutate() {
+        let chaos = Chaos {};
+	let mut rng = rand::thread_rng();
+
+        let initial_sequence: Vec<Base> = (0..10)
+	    .map(|_| rng.gen::<Base>())
+	    .collect();
+	println!("Initial sequence: ");
+	for base in &initial_sequence {
+	    print!("{}", base);
+	}
+	println!();
+	
+        let modified_sequence = chaos.full_chaos(&initial_sequence, 0.3, 0.5);
+	println!("Modified sequence: ");
+	for base in &modified_sequence {
+	    print!("{}", base);
+	}
+	println!();
+    }
+}
```

Comment:
Last ask before merge; sorry I just thought of this. Can you make some sanity check tests?

1. $P_{edit} = 0$, so no matter what $P_{mut}$ is, the sequence that is returned is not mutated, in other words, the output sequence is the exact same as the input sequence.
2. $P_{mut} = 0$, the sequence that is returned is either the exact same or has bases deleted (because $P_{del} = 1$ (length must be shorter).
3. $P_{mut} = 1$, the sequence that is returned either has exact same length as the original sequence (because $P_{del} = 0$.

Thank you for working on this!

Reply:
Hi! With my latest commit I added 3 tests and additional in-code documentation.

I'm not sure exactly how to check 2; instead, I ran it with $P_{del}$ as 1 and $P_{edit}$ as 1 so that all bases are deleted, and checked that the length is 0. Let me know if there is a better way to test for what you are hoping for.

Please let me know if there are better naming conventions I can use; I'm not particularly enamored with my current function names (esp. for tests) so let me know if we want to use some standard naming convention for all tests.

Let me know if there's anything else I can add!

Reply:
This works! Thanks! 😄

Reply:
Great! Should I merge?

Reply:
I replied on email but not sure if you got it; yep you can merge!